i18n-helpers/src/bin/mdbook-i18n.rs

```diff
@@ -0,0 +1,107 @@
+use mdbook::renderer::RenderContext;
+use serde::Deserialize;
+use std::collections::BTreeMap;
+use std::io;
+
+/// Parameters for the i18n renderer.
+///
+/// They are set in the `output.i18n` section of the book's `book.toml` file.
+#[derive(Deserialize)]
+struct I18nConfiguration {
+    /// A map of language codes to language names.
+    ///
+    /// ## Example
+    ///
+    /// ```toml
+    /// [output.i18n.languages]
+    /// "en" = "English"
+    /// "es" = "Spanish (Español)"
+    /// "ko" = "Korean (한국어)"
+    /// "pt-BR" = "Brazilian Portuguese (Português do Brasil)"
+    /// ```
+    #[serde(default)]
+    languages: BTreeMap<String, String>,
+    /// Default language code. It will not be translated.
+    default_language: Option<String>,
+
+    /// Whether to translate all languages or just the selected language, defaults to false.
+    #[serde(default)]
+    translate_all_languages: bool,
+    /// Whether to move the translations to the html directory, defaults to false.
+    ///
+    /// By default, translations' output will live in `book/i18n/<language>/<renderer>`.
+    /// For all renderers in this list, we will move individual translations to `book/<renderer>/<language>`.
+    #[serde(default)]
+    move_translations_directories: Vec<String>,
```

Comment:
It's very convenient with `mdbook serve` to view the output like it will be on the final site. It doesn't need to be enabled otherwise and by default does nothing

Reply:
>> I think you're saying here that you'll [_break the contract_ defined by mdbook](https://rust-lang.github.io/mdBook/for_developers/backends.html#output-and-signalling-failure).
>> 
>> I don't want the new renderer to touch files in the output directories of other renderers.
>
> It's very convenient with `mdbook serve` to view the output like it will be on the final site. It doesn't need to be enabled otherwise and by default does nothing

I don't think you addressed my concern: if the code is writing outside of the output directly, then the code is doing something wrong.

I don't doubt that it is convenient: we should just write everything inside of the designated output directory.

Maybe you're not aware, but I often use `mdbook build -d /tmp/foo` and similar to generate output in non-standard directories. So this renderer must respect the output setting.

Reply:
By default, we don't break the contract but if we enable this, we do. That's also what we're already doing in comprehensive rust, so I don't really see a problem as it is opt-in only.

Reply:
> That's also what we're already doing in comprehensive rust, so I don't really see a problem as it is opt-in only.

I hope our `mdbook` plugins are not breaking this?

Are you talking about [the code in `publish.yml`](https://github.com/google/comprehensive-rust/blob/77ab1ab6c41b03a62a09f22602da790b2af14717/.github/workflows/publish.yml#L60-L68), then that is quite different: the code there relies on the _contract of `mdbook`_ which says that it will put things into directory so-and-so. After calling `mdbook build`, we are of course free to move things around and mess with the `book/` directory —  we own it then!

This is the main difference between implementing this functionality as a wrapper script (as suggested in #18!) and writing it as a `mdbook` plugin. As a plugin, you are subjected to the rules of the surrounding system.

It might be a subtle difference, but I believe it's an important one. By following the principle carefully on every level of the stack, we will be able to build bigger "stacks" of software which behaves in a predictable manner when seen as a whole.

Reply:
> The script does this

Sorry, I don't think I know what script you're referring to here?

Reply:
This renderer I mean